### PR TITLE
Make the `mocks` feature work with `no_std`

### LINKED ISSUE #311: Make the code under mocks features work in no-std
## Acceptance Criteria

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate milestone (priority) applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

### PR DESCRIPTION
**Pull request by DaviRain-Su**
Closes: #311 

## Description

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests.
- [ ] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

**Comment by plafer** [comment]
Thank you for identifying this issue! However I don't think this is the right approach; it seems like we can make the entire `mocks` feature `no_std`. Namely, we are [using a `Mutex`](https://github.com/cosmos/ibc-rs/blob/060d80fc66c1de1a82c64ac193f7329ed3f2d326/crates/ibc/src/mock/context.rs#L13), which we really don't need. I'll look into it this week, and if this is indeed the case, we'll close this PR in favor of the other one I'll open.

**Comment by DaviRain-Su** [comment]
> Thank you for identifying this issue! However I don't think this is the right approach; it seems like we can make the entire `mocks` feature `no_std`. Namely, we are [using a `Mutex`](https://github.com/cosmos/ibc-rs/blob/060d80fc66c1de1a82c64ac193f7329ed3f2d326/crates/ibc/src/mock/context.rs#L13), which we really don't need. I'll look into it this week, and if this is indeed the case, we'll close this PR in favor of the other one I'll open.

`Mutex` does not fit in `no_std`, I found a crate `parking_lot` that provides `Mutex` in `no_std`, but the most important issue is `Timestamp`, which cannot be used in `no_std`, because the `now` method of `Timestamp` can only be used in `std`.

**Comment by plafer** [comment]
> but the most important issue is Timestamp, which cannot be used in no_std, because the now method of Timestamp can only be used in std.

For this, we can do solve the problem in a similar way that you did, except conditioning on the `std` feature instead of `mock-primitives`.

So I think the end solution should look like:
1. Remove the unneeded use of `Mutex`
2. Use `Timestamp::none` if the `std` feature is disabled

File: crates/ibc/Cargo.toml [hunk 1]
```diff
@@ -73,6 +73,8 @@ parity-scale-codec = { version = "3.0.0", default-features = false, features = [
 scale-info = { version = "2.1.2", default-features = false, features = ["derive"], optional = true }
 ## for borsh encode or decode
 borsh = {version = "0.9.3", default-features = false, optional = true }
+parking_lot = { version = "0.12.1", default-features = false }
+cfg-if = "1.0.0"
```
Comment (plafer) [inline crates/ibc/Cargo.toml]: Could you move those under `[dev-dependencies]`?

**Review by plafer — CHANGES_REQUESTED** [review]
Thank you for this!

Could you also add a changelog entry under "improvements"?

**Comment by plafer** [comment]
While we're at it, we should also modify the CI. In `ci/no-std-check/Cargo.toml`, could you please change line 8 from

```toml
ibc = { path = "../../crates/ibc", default-features = false }
```
to
```toml
ibc = { path = "../../crates/ibc", default-features = false, features = ["mocks"] }
```

**Comment by DaviRain-Su** [comment]
The `mock` feature also relies on `mocks = ["tendermint-testgen", "clock"]`, since `tendermint-testgen` and `clock` seem to be available only in `std`. I think it is difficult to modify it to support `no_std`, here I think it is better to expose only `Clientstate, ConsensusState, Header` in `no_std` through the feature, the whole mocks support `no_std` in the modification is still found to be very problematic.

File: crates/ibc/Cargo.toml (hunk 1 above)
Comment (DaviRain-Su) [reply crates/ibc/Cargo.toml]: I don't think it should be transferred over, because opening the mocks feature brings in some needed code

**Comment by DaviRain-Su** [comment]
I configured `tendermint-testgen` to `no-std-check` and found that `tendemrint-testgen` does not pass the `check-substrate` test.

File: crates/ibc/Cargo.toml (hunk 1 above)
Comment (plafer) [reply crates/ibc/Cargo.toml]: Ah yes you're right

**Comment by plafer** [comment]
Agreed. Let's reintroduce the feature; can we call it `mocks-no-std` instead?

I opened #329 too, which we can take care of later.

File: crates/ibc/Cargo.toml (hunk 1 above)
Comment (plafer) [reply crates/ibc/Cargo.toml]: Then can you make them optional, and a dependency of `mocks`?

**Comment by plafer** [comment]
~~`cargo test --no-default-features --features mocks-no-std` currently fails. Could you add this to the CI (in `.github/workflows/rust.yml`) as well?~~

This doesn't work on `main` either, let's not fix this here

File: crates/ibc/src/mock/header.rs [hunk 2]
```diff
@@ -71,9 +71,18 @@ impl MockHeader {
...
-        Self {
-            height,
-            timestamp: Timestamp::now(),
+        cfg_if::cfg_if! {
+            if #[cfg(any(test, feature = "mocks"))]  {
+                Self {
+                    height,
+                    timestamp: Timestamp::now(),
+                }
+            } else if #[cfg(any(test,feature = "mocks-no-std"))] {
```
Comment (plafer) [inline crates/ibc/src/mock/header.rs]:
```suggestion
            } else {
```
Improves readability, and ensures that we always compile something in.

File: crates/ibc/src/mock/header.rs [hunk 3]
```diff
@@ -71,9 +71,18 @@ impl MockHeader {
     }
 
     pub fn new(height: Height) -> Self {
-        Self {
-            height,
-            timestamp: Timestamp::now(),
+        cfg_if::cfg_if! {
+            if #[cfg(any(test, feature = "mocks"))]  {
```
Comment (plafer) [inline crates/ibc/src/mock/header.rs]:
```suggestion
            if #[cfg(any(test, feature = "std"))]  {
```
This file already only gets compiled only if `any(test, feature = "mocks", feature = "mocks-no-std")`. This better represents when we want to use `Timestamp::now()` vs `Timestamp::none()`

**Review by plafer — CHANGES_REQUESTED** [review]
nits

**Comment by DaviRain-Su** [comment]
> ~`cargo test --no-default-features --features mocks-no-std` currently fails. Could you add this to the CI (in `.github/workflows/rust.yml`) as well?~
> 
> This doesn't work on `main` either, let's not fix this here

```rust
pub fn now() -> Timestamp {
        cfg_if::cfg_if! {
            if #[cfg(feature = "clock")]  {
                Time::now().into()
            } else {
                Self::none()
            }
        }
    }
```
Try to make `Timestamp's now` method like this, but some tests will not run through.

**Review by plafer — APPROVED** [review]
Thank you!

I'll remove the `clock` feature in another PR

**Pull request merged**